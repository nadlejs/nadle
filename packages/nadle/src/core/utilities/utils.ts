import Path from "node:path";

import c from "tinyrainbow";

import { DOT } from "./constants.js";

export function noop() {}

export function capitalize(str: string): string {
	if (str.length === 0) {
		return str;
	}

	return str.charAt(0).toUpperCase() + str.slice(1);
}

const MILLISECONDS_PER_SECOND = 1000;
const MILLISECONDS_PER_MINUTE = 60 * MILLISECONDS_PER_SECOND;
const MILLISECONDS_PER_HOUR = 60 * MILLISECONDS_PER_MINUTE;

export function formatTime(ms: number): string {
	const hours = Math.floor(ms / MILLISECONDS_PER_HOUR);
	const minutes = Math.floor((ms % MILLISECONDS_PER_HOUR) / MILLISECONDS_PER_MINUTE);
	const seconds = Math.floor((ms % MILLISECONDS_PER_MINUTE) / MILLISECONDS_PER_SECOND);
	const milliseconds = Math.floor(ms % MILLISECONDS_PER_SECOND);

	if (hours > 0) {
		if (minutes > 0) {
			return `${hours}h${minutes}m`;
		}

		return `${hours}h`;
	}

	if (minutes > 0) {
		if (seconds > 0) {
			return `${minutes}m${seconds}s`;
		}

		return `${minutes}m`;
	}

	if (seconds > 0) {
		return `${seconds}s`;
	}

	return `${milliseconds}ms`;
}

export function normalizeGlobPath(path: string) {
	if (path.startsWith(DOT)) {
		return path;
	}

	return `.${Path.sep}${path}`;
}

export function clamp(value: number, min: number, max: number): number {
	return Math.min(Math.max(value, min), max);
}

export function bindObject<T, K extends keyof T>(obj: T, keys: K[]): { [P in K]: T[P] extends (...args: any[]) => any ? T[P] : never } {
	return Object.fromEntries(
		keys.map((key) => {
			const value = obj[key];

			if (typeof value !== "function") {
				throw new Error(`Property ${String(key)} is not a function`);
			}

			return [key, value.bind(obj)];
		})
	) as any;
}

export function highlight(text: string): string {
	return c.bold(c.yellow(text));
}
