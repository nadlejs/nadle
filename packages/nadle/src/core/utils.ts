export function capitalize(str: string): string {
	if (str.length === 0) {
		return str;
	}

	return str.charAt(0).toUpperCase() + str.slice(1);
}

export function formatTime(time: number): string {
	if (time > 1000) {
		return `${(time / 1000).toFixed(2)}s`;
	}

	return `${Math.round(time)}ms`;
}

export function formatTimeString(date: Date): string {
	return date.toTimeString().split(" ")[0];
}
