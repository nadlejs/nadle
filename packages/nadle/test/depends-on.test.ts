import { it, expect, describe } from "vitest";

import { getStdout, createExec } from "./setup/utils.js";

describe(
	"dependsOn",
	() => {
		const exec = createExec({ config: "depends-on" });

		it("should run dependent tasks first 1", async () => {
			const stdout = await getStdout(exec`compile`);
			expect(stdout).toRunInOrder("node", "install", ["compileSvg", "compileTs"], "compile");
		});

		it("should run dependent tasks first 2", async () => {
			const stdout = await getStdout(exec`compile test`);
			expect(stdout).toRunInOrder("node", "install", ["compileSvg", "compileTs"], "compile");
			expect(stdout).toRunInOrder("install", "test");
		});

		it("should run dependent tasks first 3", async () => {
			const stdout = await getStdout(exec`test compile`);
			expect(stdout).toRunInOrder("node", "install", ["compileSvg", "compileTs"], "compile");
			expect(stdout).toRunInOrder("install", "test");
		});

		it("should run dependent tasks first 3", async () => {
			const stdout = await getStdout(exec`build`);
			expect(stdout).toRunInOrder("node", "install", ["compileSvg", "compileTs"], "compile");
			expect(stdout).toRunInOrder("install", "test");
			expect(stdout).toRunInOrder(["test", "compile"], "build");
		});
	},
	{ retry: 0, repeats: 3 }
);
