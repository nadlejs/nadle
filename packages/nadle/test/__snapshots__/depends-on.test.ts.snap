// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dependsOn > should run dependent tasks first 1`] = `
[log] <Yellow>></Yellow> Task <Bold>node</BoldDim> started
Setup node...
[log] <Green>✓</Green> Task <Bold>node</BoldDim> done in _ms
[log] <Yellow>></Yellow> Task <Bold>install</BoldDim> started
Installing npm...
[log] <Green>✓</Green> Task <Bold>install</BoldDim> done in _ms
[log] <Yellow>></Yellow> Task <Bold>compileTs</BoldDim> started
Compiling ts...
[log] <Green>✓</Green> Task <Bold>compileTs</BoldDim> done in _ms
[log] 
<Bold><Green>RUN SUCCESSFUL</Green></BoldDim> in <Bold>_ms</BoldDim> <Dim>(<Bold>3</BoldDim><Dim> tasks executed)</BoldDim>
`;

exports[`dependsOn > should run shared dependent tasks 1`] = `
[log] <Yellow>></Yellow> Task <Bold>node</BoldDim> started
Setup node...
[log] <Green>✓</Green> Task <Bold>node</BoldDim> done in _ms
[log] <Yellow>></Yellow> Task <Bold>install</BoldDim> started
Installing npm...
[log] <Green>✓</Green> Task <Bold>install</BoldDim> done in _ms
[log] <Yellow>></Yellow> Task <Bold>compileSvg</BoldDim> started
Compiling svg...
[log] <Green>✓</Green> Task <Bold>compileSvg</BoldDim> done in _ms
[log] <Yellow>></Yellow> Task <Bold>compileTs</BoldDim> started
Compiling ts...
[log] <Green>✓</Green> Task <Bold>compileTs</BoldDim> done in _ms
[log] <Yellow>></Yellow> Task <Bold>compile</BoldDim> started
Compiling...
[log] <Green>✓</Green> Task <Bold>compile</BoldDim> done in _ms
[log] <Yellow>></Yellow> Task <Bold>test</BoldDim> started
Running tests...
[log] <Green>✓</Green> Task <Bold>test</BoldDim> done in _ms
[log] 
<Bold><Green>RUN SUCCESSFUL</Green></BoldDim> in <Bold>_ms</BoldDim> <Dim>(<Bold>6</BoldDim><Dim> tasks executed)</BoldDim>
`;

exports[`dependsOn > should run shared dependent tasks 2 1`] = `
[log] <Yellow>></Yellow> Task <Bold>node</BoldDim> started
Setup node...
[log] <Green>✓</Green> Task <Bold>node</BoldDim> done in _ms
[log] <Yellow>></Yellow> Task <Bold>install</BoldDim> started
Installing npm...
[log] <Green>✓</Green> Task <Bold>install</BoldDim> done in _ms
[log] <Yellow>></Yellow> Task <Bold>test</BoldDim> started
Running tests...
[log] <Green>✓</Green> Task <Bold>test</BoldDim> done in _ms
[log] <Yellow>></Yellow> Task <Bold>compileSvg</BoldDim> started
Compiling svg...
[log] <Green>✓</Green> Task <Bold>compileSvg</BoldDim> done in _ms
[log] <Yellow>></Yellow> Task <Bold>compileTs</BoldDim> started
Compiling ts...
[log] <Green>✓</Green> Task <Bold>compileTs</BoldDim> done in _ms
[log] <Yellow>></Yellow> Task <Bold>compile</BoldDim> started
Compiling...
[log] <Green>✓</Green> Task <Bold>compile</BoldDim> done in _ms
[log] 
<Bold><Green>RUN SUCCESSFUL</Green></BoldDim> in <Bold>_ms</BoldDim> <Dim>(<Bold>6</BoldDim><Dim> tasks executed)</BoldDim>
`;
